### PR TITLE
Fuel

### PR DESCRIPTION
**Pull request by duzos**
This will add fuel to Tardises

TODO:
- [x] Drain fuel in places
- [ ] ~~Add proper method for refueling Tardis ( ideas needed )~~
- [ ] ~~Show amount of fuel a Tardis has~~ ( Unsure how to use `MonitorScreen` )
- [x] Improve out of fuel experience

### CODE REVIEW

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java [hunk 1]
```diff
@@ -35,7 +36,8 @@
 import static whocraft.tardis_refined.common.util.TardisHelper.isInArsArea;
 
 public class TardisClientData {
-
+    public static int FOG_TICK_DELTA = 0; // This is for the fading in and out of the fog.
+    private static int MAX_FOG_TICK_DELTA = 2 * 20; // This is for adjusting how fast the fog will fade in and out.
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java:40]: Very nice!

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java [hunk 2]
```diff
@@ -61,6 +63,8 @@ public ResourceKey<Level> getLevelKey() {
     private boolean isCrashing = false;
     private boolean isOnCooldown = false;
     private float flightShakeScale = 0;
+    private double fuel = 0;
+    private double maximumFuel = 0;
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java:67]: Very nice!

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java [hunk 3]
```diff
@@ -161,6 +165,38 @@ public float flightShakeScale() {
...
+    public double getMaximumFuel() {
+        return maximumFuel;
+    }
+    public void setMaximumFuel(double fuel) {
+        this.maximumFuel = fuel;
+    }
+
+    /**
+     * Higher means more fog, lower means less fog
+     * @return 0 -> 1 float based off fog tick delta
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java:183]: Very nice! Love that's documented!

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java [hunk 4]
```diff
@@ -182,6 +218,9 @@ public CompoundTag serializeNBT() {
 
         compoundTag.putString(NbtConstants.TARDIS_CURRENT_HUM, humEntry.getIdentifier().toString());
 
+        compoundTag.putDouble(NbtConstants.FUEL, fuel);
+        compoundTag.putDouble(NbtConstants.MAXIMUM_FUEL, maximumFuel);
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java:222]: Good good, I like that we're using NBTConstants

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java [hunk 5]
```diff
@@ -255,6 +297,11 @@ public void tickClientside() {
                 HumSoundManager.playHum(SoundEvent.createVariableRangeEvent(humEntry.getSound()));
             }
 
+            // Stops the hum sound if there's no fuel
+            if (isThisTardis && fuel == 0 && soundManager.isActive(HumSoundManager.getCurrentSound())) {
+                HumSoundManager.playHum(SoundEvents.EMPTY);
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java]: We usually just set the volume to 0 so there's less stopping and starting of the sound system

File: common/src/main/java/whocraft/tardis_refined/common/capability/TardisLevelOperator.java [hunk 6]
```diff
@@ -142,6 +142,8 @@ public void tick(ServerLevel level) {
             tardisClientData.setShellTheme(aestheticHandler.getShellTheme());
             tardisClientData.setShellPattern(aestheticHandler.shellPattern().id());
             tardisClientData.setHumEntry(interiorManager.getHumEntry());
+            tardisClientData.setFuel(pilotingManager.getFuel());
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/capability/TardisLevelOperator.java:145]: I cannot wait for TardisClientData to die

File: common/src/main/java/whocraft/tardis_refined/common/mixin/FogRendererMixin.java [hunk 7]
```diff
@@ -29,12 +35,29 @@ private static void setupFog(Camera camera, FogRenderer.FogMode fogMode, float f
...
 
-    @Inject(at = @At("HEAD"), cancellable = true, method = "levelFogColor()V")
+    @Inject(at = @At("HEAD"), cancellable = true, method = "setupFog")
     private static void setupColor(CallbackInfo callbackInfo) {
         if (Minecraft.getInstance().player != null) {
             BlockPos blockPosition = Minecraft.getInstance().player.blockPosition();
+
+            ClientLevel level = Minecraft.getInstance().level;
+
+            if (level.dimensionTypeId() != DimensionTypes.TARDIS) return;
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/mixin/FogRendererMixin.java:45]: Very good

File: common/src/main/java/whocraft/tardis_refined/common/network/messages/ChangeDesktopMessage.java [hunk 8]
```diff
@@ -52,9 +53,11 @@ public void handle(MessageContext context) {
         level.ifPresent(x -> {
             TardisLevelOperator.get(x).ifPresent(operator -> {
                 TardisPilotingManager pilotManager = operator.getPilotingManager();
+                TardisInteriorManager interiorManager = operator.getInteriorManager();
 
-                if (!pilotManager.isInFlight()) {
-                    operator.getInteriorManager().prepareDesktop(desktopTheme);
+                if (!pilotManager.isInFlight() && interiorManager.hasEnoughFuel()) {
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/network/messages/ChangeDesktopMessage.java]: We should maybe play a sound when the desktop cannot be changed?

File: common/src/main/java/whocraft/tardis_refined/common/tardis/control/ship/MonitorControl.java [hunk 9]
```diff
@@ -16,10 +19,19 @@ public class MonitorControl extends Control {
...
+                operator.getPilotingManager().setFuel(operator.getPilotingManager().getMaximumFuel());
+                player.sendSystemMessage(Component.literal("Fueled up!"));
+                return true;
+            }
+
             boolean isSyncingKey = false;
             if (PlayerUtil.isInMainHand(player, ItemRegistry.KEY.get())){
-                KeyItem key = (KeyItem)player.getMainHandItem().getItem();
-                if (key.interactMonitor(player.getMainHandItem(),player, controlEntity, player.getUsedItemHand()))
+                KeyItem key = (KeyItem)hand.getItem();
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/control/ship/MonitorControl.java]: no! bad duzo! PlayerUtil kinda sucks for this kinda thing
Instead use use pattern matching https://docs.oracle.com/en/java/javase/17/language/pattern-matching-instanceof-operator.html#GUID-843060B5-240C-4F47-A7B0-95C42E5B08A7

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/AestheticHandler.java [hunk 10]
```diff
@@ -32,6 +32,7 @@ public class AestheticHandler extends BaseHandler {
     private ResourceLocation shellTheme = ShellTheme.FACTORY.getId();
     private ShellPattern shellPattern = ShellPatterns.DEFAULT;
     private HumEntry currentHum = TardisHums.getDefaultHum();
+    public static final double REQUIRED_SHELL_FUEL = 15; // Amount of fuel required to change the shell todo - make non static
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/AestheticHandler.java]: Mixed feelings about this being here

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisFlightEventManager.java [hunk 11]
```diff
@@ -195,15 +195,17 @@ public int getBlocksPerRequest(double distance) {
...
+        if (pilot.isInFlight() && !pilot.isAutoLandSet()) {
+
+            if (!pilot.isCrashing()) {
                 ticksSincePrompted++;
 
                 if (controlRequestCooldown > 0) controlRequestCooldown--;
 
                 // Prepare the next control for highlighting.
-                if (!isWaitingForControlResponse && controlRequestCooldown == 0 && this.controlResponses < this.requiredControlRequests && !operator.getPilotingManager().isAutoLandSet()) {
+                if (!isWaitingForControlResponse && controlRequestCooldown == 0 && this.controlResponses < this.requiredControlRequests && !pilot.isAutoLandSet()) { // "&& !pilot.isAutoLandSet()" should not be necessary as it was checked above
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisFlightEventManager.java]: Remove it

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisFlightEventManager.java [hunk 12]
```diff
@@ -235,7 +237,7 @@ public void tick() {
 
                 }
 
-                if (this.isInDangerZone) {
+                if (this.isInDangerZone || pilot.isOutOfFuel()) {
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisFlightEventManager.java:240]: Oooo

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisInteriorManager.java [hunk 13]
```diff
@@ -53,6 +49,8 @@ public class TardisInteriorManager extends BaseHandler {
 
     public static final BlockPos STATIC_CORRIDOR_POSITION = new BlockPos(1000, 100, 0);
 
+    public static final double REQUIRED_FUEL_INT_CHANGE = 500; // The amount of fuel required to change interior todo - move into non static so can be changed by upgrades and such
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisInteriorManager.java]: get set

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisPilotingManager.java [hunk 14]
```diff
@@ -150,10 +169,14 @@ public void tick(Level level) {
             }
         }
 
-
         if (isInFlight) {
             ticksInFlight++;
 
+            // Removing fuel once every 2.5 seconds, should this be here?
+            if (ticksInFlight % (2.5 * 20) == 0) {
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisPilotingManager.java:176]: Looks fine to be here

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisPilotingManager.java [hunk 15]
```diff
@@ -193,7 +216,15 @@ public void tick(Level level) {
                 this.operator.getLevel().playSound(null, TardisArchitectureHandler.DESKTOP_CENTER_POS, SoundRegistry.TARDIS_SINGLE_FLY.get(), SoundSource.AMBIENT, 100f, 0.25f);
             }
         }
-
+        /*
+        if (this.isOutOfFuel()) {
+            // i have concerns about this causing lag, lmk if it does and ill rework it
```
Comment (Jeryn99) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisPilotingManager.java]: @CommandrMoose's opinion on this would be best

**Review by Jeryn99 — CHANGES_REQUESTED** [review]
Looks good, just some small changes to make

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java (hunk 5 above)
Comment (duzos) [reply common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java]: Will change
Is there a method or something in the hum manager?

File: common/src/main/java/whocraft/tardis_refined/common/capability/TardisLevelOperator.java (hunk 6 above)
Comment (duzos) [reply common/src/main/java/whocraft/tardis_refined/common/capability/TardisLevelOperator.java:145]: Agreed

File: common/src/main/java/whocraft/tardis_refined/common/network/messages/ChangeDesktopMessage.java (hunk 8 above)
Comment (duzos) [reply common/src/main/java/whocraft/tardis_refined/common/network/messages/ChangeDesktopMessage.java]: Good point i'll add some temporary nuh uh sound

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisPilotingManager.java (hunk 15 above)
Comment (duzos) [reply common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisPilotingManager.java]: Forgot to remove that, isn't needed anymore cus of the fog mixin

File: common/src/main/java/whocraft/tardis_refined/common/tardis/control/ship/MonitorControl.java (hunk 9 above)
Comment (duzos) [reply common/src/main/java/whocraft/tardis_refined/common/tardis/control/ship/MonitorControl.java]: this wasnt meeee but ill change it anyways

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/AestheticHandler.java (hunk 10 above)
Comment (duzos) [reply common/src/main/java/whocraft/tardis_refined/common/tardis/manager/AestheticHandler.java]: unsure where it should go but for now ill move it to `TardisExteriorManager` makes a smidge more sense i guess

File: common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java (hunk 5 above)
Comment (Jeryn99) [reply common/src/main/java/whocraft/tardis_refined/client/TardisClientData.java]: @Duzos No, just inside loopingsound itself

**Pull request merged**